framework/bundles/org.jboss.tools.rsp.server/src/main/java/org/jboss/tools/rsp/server/filewatcher/FileWatcherService.java

```diff
@@ -0,0 +1,584 @@
+		
+		// A request for a subfolder (or lower) still exists, so I'm still needed
+		Set<Path> requestPaths = requests.keySet();
+		for( Path testRequest : requestPaths) {
+			if( testRequest.startsWith(path))
+				return true;
+		}
+
+		// Nobody needs me
+		return false;
+	}
+	
+	private boolean pathOrParentHasRecursiveRequest(Path p) {
+		Path working = p;
+		while(working != null ) {
+			if( singlePathHasRecursiveRequest(working)) {
+				return true;
+			}
+			working = working.getParent();
+		}
+		return false;
+	}
+	private boolean singlePathHasRecursiveRequest(Path p) {
+		List<RegistrationRequest> list = requests.get(p);
+		if( list == null ) 
+			return false;
+		Iterator<RegistrationRequest> rit = list.iterator();
+		RegistrationRequest r = null;
+		while(rit.hasNext()) {
+			r = rit.next();
+			if( r.isRecursive())
+				return true;
+		}
+		return false;
+	}
+
+	@Override
+	public void run() {
+		WatchKey key;
+		try {
+			while (watchService != null && (key = watchService.take()) != null) {
+				// Get the list and reset immedietely
+				List<WatchEvent<?>> events = key.pollEvents();
+				key.reset();
+				
+				for (WatchEvent<?> event : events) {
+					Path eventContext = (Path)event.context();
+					if( eventContext != null ) {
+						Path context = ((Path)key.watchable()).resolve((Path)event.context());
+						subscribeToChanges(event, context);
+						fireEvents(key, event);
+					}
+				}
+			}
+		} catch (InterruptedException | ClosedWatchServiceException e) {
+			if( isClosing() == false ) {
+				// unexpected error
+				e.printStackTrace();
+			}
+		}
+		setClosing(false);
+	}
+
+	/*
+	 * Subscribe to any newly created folders that are below a recursive 
+	 * subscription. Cleanup any watch keys that are already listening 
+	 * to a folder that may have been deleted.  In short, make 
+	 * sure the watch service is listening to the right things after
+	 * whatever changes have been detected. 
+	 */
+	private synchronized void subscribeToChanges(WatchEvent<?> event, Path eventContext) {
+		if( event.kind() == StandardWatchEventKinds.ENTRY_DELETE) {
+			// something we were watching has been deleted. Let's 
+			// make sure that we delete all current watch keys for that 
+			// and any sub-directories if possible
+			removeAllSubscriptionsRecursive(eventContext);
+		} else if( event.kind() == StandardWatchEventKinds.ENTRY_CREATE) {
+			if( eventContext.toFile().isDirectory()) {
+				// a new folder has been created. 
+				// Let's see if any requests match this path, 
+				// or if any recursive requests match our parents
+				boolean recursive = recursiveRequestMatches(eventContext);
+				if( recursive ) {
+					ensureChildrenSubscribed(eventContext);
+				} else if( requestMatchesExact(eventContext) ) {
+					subscribeSinglePath(eventContext);
+				}
+			}
+		}
+		// ignore modify event kind
+		
+	}
+
+	private boolean requestMatchesExact(Path path) {
+		return requests.get(path) != null && requests.get(path).size() > 0;
```

Comment:
```!requests.get(path).isEmpty();```

Reply:
Not sure why you focus on this so much ;)   ArrayList.isEmpty() just returns size == 0   lol.   So it's the same thing.